packages/query-core/src/queryObserver.ts

```diff
@@ -593,6 +608,42 @@ export class QueryObserver<
       | undefined
 
     const nextResult = this.createResult(this.#currentQuery, this.options)
+
+    if (this.options.experimental_prefetchInRender) {
+      const nextThenable = (() => {
+        const thenable = pendingThenable<TData>()
+
+        if (nextResult.status === 'error') {
+          thenable.reject(nextResult.error)
+        } else if (nextResult.data !== undefined) {
+          thenable.resolve(nextResult.data)
+        } else if (!nextResult.isFetching) {
+          thenable.reject(cancellationError)
+        }
+        return thenable as Thenable<TData>
+      })()
+
+      const prevThenable = this.#currentThenable
+
+      switch (prevThenable.status) {
+        case 'pending':
+          if (nextThenable.status === 'fulfilled') {
+            prevThenable.resolve(nextThenable.value)
+          } else if (nextThenable.status === 'rejected') {
+            prevThenable.reject(nextThenable.reason)
+          }
+          break
+
+        case 'fulfilled':
+        case 'rejected':
+          if (!isThenableEqual(prevThenable, nextThenable)) {
+            // Replace the thenable when the results have changed
+            this.#currentThenable = nextResult.promise = nextThenable
+          }
+          break
+      }
+    }
```

Comment:
cc @TkDodo 

This code has updated a bit to support for cancellation - since the query client itself doesn't have the concept of "abort errors" (AFAICT?) we instead:

1. Eagerly create a new thenable of what the "next" thenable should represent
  1. If the previous thenable is pending, we finalize it if the next reason or value
  2. If the previous is not pending, we do an equal-check to see if we should replace it (since the cancellation error is a constant, the equal-check will work)

Reply:
@KATT I think I changed my mind on this one. Rejecting the promise doesn't seem right - the user shouldn't see such a cancelation error, as it's not actionable. I think we can just let the promise stay pending, and then it needs to resolve when data comes in.

I did that here: [849b1b6](https://github.com/TanStack/query/pull/7988/commits/849b1b6d22c679fe9f522ee523a5bd76ca00c0f1)

let me know if that's okay, implementation wise?

Reply:
it's pretty much the same behaviour if a query is `enabled: false` and you call `use()` on the promise. Basically every situation where you are `status: 'pending'` and `fetchStatus` isn't `'fetching'`. Added a test for that here: [c75afaa](https://github.com/TanStack/query/pull/7988/commits/c75afaaae9a266447a7a805326fdae01a7b0cde2)

Reply:
Sounds fine to me, does that align with the behavior of

```ts
const resPromise = client.fetchQuery(/* ... */)
client.cancelQueries(/* ... */)

const res = await resPromise; // <-- never resolves currently, i guess?
```

it does seem to be that this should result in an [`AbortError`](https://developer.mozilla.org/en-US/docs/Web/API/DOMException#aborterror) but if it aligns with the current behavior of a fetch/cancel dance, it sgtm